In [1]:
import pandas as pd
import numpy as np
from impute_df import impute_missing_values

In [2]:
# Load the data from the csv file
df = pd.read_csv('Protenus Sales Training Data_Update.csv')
df.head()

Snapshot_Date   Opportunity_ID      Opportunity_Stage   Opportunity_Type  \
0        7/1/23  0061K00000mR3oA   Stage 1 - Evaluating       New Business   
1        7/1/23  0061K00000lUGm2  Stage 0 - Prospecting  Existing Business   
2        7/1/23  0061K00000lUCBW      Stage 2 - Scoping       New Business   
3        7/1/23  0061K00000lUBat  Stage 0 - Prospecting       New Business   
4        7/1/23  0061K00000lU5Vh  Stage 0 - Prospecting  Existing Business   

   Opportunity_Annual_Amount Opportunity_Created_Date Opportunity_Close_Date  \
0                292073.3989                  6/30/23                6/26/24   
1                102985.0200                  6/28/23               12/15/23   
2                154282.5407                  6/26/23                6/30/24   
3                166344.3180                  6/26/23               12/29/23   
4                120951.0000                  6/21/23                 2/1/24   

    Opportunity_Product Account_Electronic_Medical_Record_System  \
0    Privacy Monitoring                                     Epic   
1  Diversion Monitoring                                     Epic   
2  Diversion Monitoring                                     Epic   
3    Privacy Monitoring                                     Epic   
4  Diversion Monitoring                                    Other   

            Account_Hospital_Type Account_Fiscal_Year_End Account_State  \
0                   Health System                     NaN            VA   
1  Short Term Acute Care Hospital                 6/30/23            CA   
2                   Health System                     NaN            VA   
3  Short Term Acute Care Hospital                 9/30/23            TX   
4                             NaN                     NaN           NaN   

   Account_Zip_Code  Account_Number_of_Hospitals   Account_Number_of_Beds  \
0           24016.0                           8.0                   951.0   
1           92868.0                          24.0                   379.0   
2           23298.0                           9.0                  1942.0   
3           76104.0                           0.0                   434.0   
4               NaN                           NaN                     NaN   

   Account_Number_of_Affiliated_Physicians  Account_Number_of_Employee  \
0                                   1522.0                         NaN   
1                                   1143.0                         NaN   
2                                   1980.0                      8571.0   
3                                    775.0                         NaN   
4                                      NaN                         NaN   

   Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
0             5228803.0                   6.217398e+09   
1                   NaN                   5.277530e+09   
2           298090050.0                   8.342886e+09   
3           131280220.0                   3.195426e+09   
4                   NaN                            NaN   

   Account_Net_Patient_Revenue  
0                 1.933876e+09  
1                 1.659711e+09  
2                 2.248396e+09  
3                 6.079643e+08  
4                          NaN

In [3]:
# Drop all the old data with missing annual amounts
df['Snapshot_Date'] = pd.to_datetime(df['Snapshot_Date'])
df = df[df['Snapshot_Date'] >= '2020-08-01']
print(df.shape)

(7542, 20)


In [4]:
df = impute_missing_values(df)
df.head()

Snapshot_Date   Opportunity_ID  Opportunity_Stage Opportunity_Type  \
5979    2021-04-01  0061K00000b3JMy        Closed Lost     New Business   
6171    2021-03-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6359    2021-02-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6551    2021-01-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6745    2020-12-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   

      Opportunity_Annual_Amount Opportunity_Created_Date  \
5979                167838.3211                   3/5/19   
6171                167838.3211                   3/5/19   
6359                167838.3211                   3/5/19   
6551                167838.3211                   3/5/19   
6745                167838.3211                   3/5/19   

     Opportunity_Close_Date Opportunity_Product  \
5979                3/30/21  Privacy Monitoring   
6171               12/30/21  Privacy Monitoring   
6359               12/30/21  Privacy Monitoring   
6551                6/30/21  Privacy Monitoring   
6745                6/30/21  Privacy Monitoring   

     Account_Electronic_Medical_Record_System           Account_Hospital_Type  \
5979                                    Other  Short Term Acute Care Hospital   
6171                                    Other  Short Term Acute Care Hospital   
6359                                    Other  Short Term Acute Care Hospital   
6551                                    Other  Short Term Acute Care Hospital   
6745                                    Other  Short Term Acute Care Hospital   

     Account_Fiscal_Year_End Account_State  Account_Zip_Code  \
5979                 6/30/23            PA           15243.0   
6171                 6/30/23            PA           15243.0   
6359                 6/30/23            PA           15243.0   
6551                 6/30/23            PA           15243.0   
6745                 6/30/23            PA           15243.0   

      Account_Number_of_Hospitals   Account_Number_of_Beds  \
5979                     15.514252                   296.0   
6171                     15.514252                   296.0   
6359                     15.514252                   296.0   
6551                     15.514252                   296.0   
6745                     15.514252                   296.0   

      Account_Number_of_Affiliated_Physicians  Account_Number_of_Employee  \
5979                                    684.0                10236.916088   
6171                                    684.0                10236.916088   
6359                                    684.0                10236.916088   
6551                                    684.0                10236.916088   
6745                                    684.0                10236.916088   

      Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
5979            21723248.0                   1.471406e+09   
6171            21723248.0                   1.471406e+09   
6359            21723248.0                   1.471406e+09   
6551            21723248.0                   1.471406e+09   
6745            21723248.0                   1.471406e+09   

      Account_Net_Patient_Revenue  
5979                  370594478.0  
6171                  370594478.0  
6359                  370594478.0  
6551                  370594478.0  
6745                  370594478.0

In [5]:
df['Month'] = df['Snapshot_Date'].dt.month - 1
df.loc[df['Snapshot_Date'].dt.month == 1, 'Month'] = 12
df['Year'] = df['Snapshot_Date'].dt.year
df.loc[df['Snapshot_Date'].dt.month == 1, 'Year'] = df['Year'] - 1

# Create a new column for quarters where 1,2,3 = Q1, 4,5,6 = Q2, 7,8,9 = Q3, 10,11,12 = Q4
df['Quarter'] = np.where(df['Month'].isin([1,2,3]), 'Q1',
                         np.where(df['Month'].isin([4,5,6]), 'Q2',
                                  np.where(df['Month'].isin([7,8,9]), 'Q3',
                                           np.where(df['Month'].isin([10,11,12]), 'Q4', 'NA'))))
df.head()

Snapshot_Date   Opportunity_ID  Opportunity_Stage Opportunity_Type  \
5979    2021-04-01  0061K00000b3JMy        Closed Lost     New Business   
6171    2021-03-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6359    2021-02-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6551    2021-01-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   
6745    2020-12-01  0061K00000b3JMy  Stage 2 - Scoping     New Business   

      Opportunity_Annual_Amount Opportunity_Created_Date  \
5979                167838.3211                   3/5/19   
6171                167838.3211                   3/5/19   
6359                167838.3211                   3/5/19   
6551                167838.3211                   3/5/19   
6745                167838.3211                   3/5/19   

     Opportunity_Close_Date Opportunity_Product  \
5979                3/30/21  Privacy Monitoring   
6171               12/30/21  Privacy Monitoring   
6359               12/30/21  Privacy Monitoring   
6551                6/30/21  Privacy Monitoring   
6745                6/30/21  Privacy Monitoring   

     Account_Electronic_Medical_Record_System           Account_Hospital_Type  \
5979                                    Other  Short Term Acute Care Hospital   
6171                                    Other  Short Term Acute Care Hospital   
6359                                    Other  Short Term Acute Care Hospital   
6551                                    Other  Short Term Acute Care Hospital   
6745                                    Other  Short Term Acute Care Hospital   

      ... Account_Number_of_Hospitals  Account_Number_of_Beds  \
5979  ...                    15.514252                  296.0   
6171  ...                    15.514252                  296.0   
6359  ...                    15.514252                  296.0   
6551  ...                    15.514252                  296.0   
6745  ...                    15.514252                  296.0   

      Account_Number_of_Affiliated_Physicians  Account_Number_of_Employee  \
5979                                    684.0                10236.916088   
6171                                    684.0                10236.916088   
6359                                    684.0                10236.916088   
6551                                    684.0                10236.916088   
6745                                    684.0                10236.916088   

      Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
5979            21723248.0                   1.471406e+09   
6171            21723248.0                   1.471406e+09   
6359            21723248.0                   1.471406e+09   
6551            21723248.0                   1.471406e+09   
6745            21723248.0                   1.471406e+09   

      Account_Net_Patient_Revenue  Month  Year  Quarter  
5979                  370594478.0      3  2021       Q1  
6171                  370594478.0      2  2021       Q1  
6359                  370594478.0      1  2021       Q1  
6551                  370594478.0     12  2020       Q4  
6745                  370594478.0     11  2020       Q4  

[5 rows x 23 columns]

In [6]:
# Create a quarterly dataframe, where for each 'Opportunity_ID' the data is aggregated by quarter and year. 
df['Year_Quarter'] = df['Year'].astype(str) + df['Quarter'].astype(str)

quarterly_df = df.sort_values('Snapshot_Date').groupby(['Opportunity_ID', 'Year_Quarter']).last().reset_index()

quarterly_df.head()

Opportunity_ID Year_Quarter Snapshot_Date  Opportunity_Stage  \
0  0061K00000b3JMy       2020Q3    2020-10-01  Stage 2 - Scoping   
1  0061K00000b3JMy       2020Q4    2021-01-01  Stage 2 - Scoping   
2  0061K00000b3JMy       2021Q1    2021-04-01        Closed Lost   
3  0061K00000b3Qsm       2020Q3    2020-10-01  Stage 3 - Pricing   
4  0061K00000b3Qsm       2020Q4    2021-01-01  Stage 3 - Pricing   

  Opportunity_Type  Opportunity_Annual_Amount Opportunity_Created_Date  \
0     New Business                167838.3211                   3/5/19   
1     New Business                167838.3211                   3/5/19   
2     New Business                167838.3211                   3/5/19   
3     New Business                418991.1971                   3/7/19   
4     New Business                418991.1971                   3/7/19   

  Opportunity_Close_Date Opportunity_Product  \
0                3/31/21  Privacy Monitoring   
1                6/30/21  Privacy Monitoring   
2                3/30/21  Privacy Monitoring   
3                4/16/21  Privacy Monitoring   
4                9/16/21  Privacy Monitoring   

  Account_Electronic_Medical_Record_System  ... Account_Number_of_Hospitals   \
0                                    Other  ...                    15.514252   
1                                    Other  ...                    15.514252   
2                                    Other  ...                    15.514252   
3                                     Epic  ...                    21.000000   
4                                     Epic  ...                    21.000000   

  Account_Number_of_Beds Account_Number_of_Affiliated_Physicians  \
0                  296.0                                   684.0   
1                  296.0                                   684.0   
2                  296.0                                   684.0   
3                 2708.0                                  3670.0   
4                 2708.0                                  3670.0   

   Account_Number_of_Employee  Account_Cash_on_Hand  \
0                10236.916088            21723248.0   
1                10236.916088            21723248.0   
2                10236.916088            21723248.0   
3                12464.000000           160091963.0   
4                12464.000000           160091963.0   

   Account_Total_Patient_Revenue  Account_Net_Patient_Revenue  Month  Year  \
0                   1.471406e+09                 3.705945e+08      9  2020   
1                   1.471406e+09                 3.705945e+08     12  2020   
2                   1.471406e+09                 3.705945e+08      3  2021   
3                   1.345325e+10                 2.802628e+09      9  2020   
4                   1.345325e+10                 2.802628e+09     12  2020   

   Quarter  
0       Q3  
1       Q4  
2       Q1  
3       Q3  
4       Q4  

[5 rows x 24 columns]

In [7]:
# Create a column for the 'Opportunity_Annual_Amount' of the previous quarter for each 'Opportunity_ID'
quarterly_df['Opportunity_Annual_Amount_Prev_Quarter'] = quarterly_df.groupby('Opportunity_ID')['Opportunity_Annual_Amount'].shift(1)
quarterly_df.head()

# Filter the data to only include data where 'Opportunity_Stage' is 'Closed Won'
quarterly_df_won = quarterly_df[quarterly_df['Opportunity_Stage'] == 'Closed Won']
quarterly_df_won.head()

# Drop data where 'Opportunity_Annual_Amount' is null or 0
# quarterly_df_won = quarterly_df_won[quarterly_df_won['Opportunity_Annual_Amount'] > 0]
# quarterly_df_won.head()


Opportunity_ID Year_Quarter Snapshot_Date Opportunity_Stage  \
12  0061K00000b3Qsm       2022Q4    2023-01-01        Closed Won   
51  0061K00000c4njV       2023Q2    2023-06-01        Closed Won   
62  0061K00000cP8i8       2020Q3    2020-10-01        Closed Won   
63  0061K00000cPETF       2020Q3    2020-08-01        Closed Won   
72  0061K00000chhPZ       2020Q3    2020-10-01        Closed Won   

     Opportunity_Type  Opportunity_Annual_Amount Opportunity_Created_Date  \
12       New Business               383667.85170                   3/7/19   
51       New Business               523803.90910                   3/1/19   
62       New Business               376921.72130                   5/1/19   
63  Existing Business                94503.85434                   5/2/19   
72       New Business               139759.99490                  3/26/19   

   Opportunity_Close_Date   Opportunity_Product  \
12                12/2/22    Privacy Monitoring   
51                5/18/23  Diversion Monitoring   
62                9/15/20  Compliance Analytics   
63                7/23/20  Diversion Monitoring   
72                9/30/20    Privacy Monitoring   

   Account_Electronic_Medical_Record_System  ... Account_Number_of_Beds  \
12                                     Epic  ...                 2708.0   
51                                     Epic  ...                 4816.0   
62                                     Epic  ...                  440.0   
63                                    Other  ...                  431.0   
72                                     Epic  ...                  575.0   

   Account_Number_of_Affiliated_Physicians Account_Number_of_Employee  \
12                              3670.00000                    12464.0   
51                              6663.00000                    23378.0   
62                              2841.00000                     5674.0   
63                              2910.37336                     2825.0   
72                              1550.00000                     7600.0   

    Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
12          1.600920e+08                   1.345325e+10   
51          1.592057e+08                   2.118838e+10   
62          1.396530e+08                   3.156241e+09   
63          2.555342e+08                   1.113518e+10   
72          3.510172e+07                   4.250111e+09   

    Account_Net_Patient_Revenue  Month  Year  Quarter  \
12                 2.802628e+09     12  2022       Q4   
51                 6.505254e+09      5  2023       Q2   
62                 1.194338e+09      9  2020       Q3   
63                 2.795598e+09      7  2020       Q3   
72                 1.201182e+09      9  2020       Q3   

    Opportunity_Annual_Amount_Prev_Quarter  
12                             396827.9728  
51                             523803.9091  
62                                     NaN  
63                                     NaN  
72                                     NaN  

[5 rows x 25 columns]

In [8]:
print(df.shape)
print(quarterly_df.shape)
print(quarterly_df_won.shape)

(7542, 24)
(2925, 25)
(142, 25)


In [9]:
# Calculate the difference between the 'Opportunity_Annual_Amount' of the current quarter and the previous quarter
quarterly_df_won['Opportunity_Annual_Amount_Diff'] = quarterly_df_won['Opportunity_Annual_Amount'] - quarterly_df_won['Opportunity_Annual_Amount_Prev_Quarter']
quarterly_df_won.head()

# Calculate the average of 'Opportunity_Annual_Amount_Diff' for all 'Opportunity_ID's
Q_diff = quarterly_df_won['Opportunity_Annual_Amount_Diff'].mean()

# Calculate the average of 'Opportunity_Annual_Amount' and get a percentage of the average difference
Q_avg = quarterly_df_won['Opportunity_Annual_Amount'].mean()

print(Q_diff)
print(Q_avg)
print (100* Q_diff/Q_avg)


-15381.217533877052
189334.54958081467
-8.123830314082113


C:\Users\odani\AppData\Local\Temp\ipykernel_24316\1376284020.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  quarterly_df_won['Opportunity_Annual_Amount_Diff'] = quarterly_df_won['Opportunity_Annual_Amount'] - quarterly_df_won['Opportunity_Annual_Amount_Prev_Quarter']


The above would mean that on average, the closing deal goes down by ~11 percent compared to the previous quarter estimation.

In [10]:
# Count the number of 'Opportunity_ID's that only appear once in the quarterly data
ids_once = quarterly_df['Opportunity_ID'].value_counts()
ids_once = ids_once[ids_once == 1].index

# Create a subset of the data where 'Opportunity_ID' only appears one time
quarter_df_once = quarterly_df[quarterly_df['Opportunity_ID'].isin(ids_once)]

# Count how many in quarter_df_once have a 'Closed Won' 'Opportunity_Stage'
stages = quarter_df_once['Opportunity_Stage'].value_counts()
Closed_singleQ = stages['Closed Won']

# Count how many 'Closed Won' 'Opportunity_ID's there are in the quarterly data
Closed_total = quarterly_df_won['Opportunity_ID'].nunique()

print(Closed_singleQ)
print(Closed_total)
print(100*Closed_singleQ/(Closed_total))


29
142
20.422535211267604


The above indicated that 29 out of the total 142 closed deals were opened and closed in the same quarter. This is ~20% of all deals.

In [11]:
# Count total number of quarters in the data
total_quarters = quarterly_df['Year_Quarter'].nunique()
print(total_quarters)

# Count the average amount of closings per quarter
avg_closings = Closed_total/total_quarters
print(avg_closings)

# Count the average annual amount per quarter
avg_annual = avg_closings * Q_avg
print(avg_annual)

# Count how many 'Closed Won' cases there were for each 'Year_Quarter'
Closed_quarters = quarterly_df_won['Year_Quarter'].value_counts()
print(Closed_quarters)


12
11.833333333333334
2240458.836706307
2021Q4    22
2022Q4    21
2020Q4    13
2023Q1    12
2021Q3    12
2023Q2    11
2020Q3    11
2022Q2    11
2022Q3    10
2021Q2     8
2021Q1     6
2022Q1     5
Name: Year_Quarter, dtype: int64


There are a total of 12 quarters, where on average there are ~12 closing per quarter.

In [12]:
# To account for 'Closed_singleQ', calculate its quarterly effect
prob_singleQ = Closed_singleQ/total_quarters
avg_singleQ = prob_singleQ * Q_avg
print(avg_singleQ)

# To account for 'Q_diff', calculate its quarterly effect
avg_Q_diff = Q_diff * avg_closings
print(avg_Q_diff)

# Calculate the total quarterly effect
total_avg_Q_eff = avg_singleQ + avg_Q_diff
print (100*total_avg_Q_eff/avg_annual)

457558.49482030206
-182011.07415087847
12.29870489718549


In [13]:
# Save the quarterly_df as a csv file
quarterly_df['Opportunity_Annual_Amount_Diff'] = quarterly_df['Opportunity_Annual_Amount'] - quarterly_df['Opportunity_Annual_Amount_Prev_Quarter']
quarterly_df.to_csv('quarterly_df.csv', index=False)
quarterly_df.head()

Opportunity_ID Year_Quarter Snapshot_Date  Opportunity_Stage  \
0  0061K00000b3JMy       2020Q3    2020-10-01  Stage 2 - Scoping   
1  0061K00000b3JMy       2020Q4    2021-01-01  Stage 2 - Scoping   
2  0061K00000b3JMy       2021Q1    2021-04-01        Closed Lost   
3  0061K00000b3Qsm       2020Q3    2020-10-01  Stage 3 - Pricing   
4  0061K00000b3Qsm       2020Q4    2021-01-01  Stage 3 - Pricing   

  Opportunity_Type  Opportunity_Annual_Amount Opportunity_Created_Date  \
0     New Business                167838.3211                   3/5/19   
1     New Business                167838.3211                   3/5/19   
2     New Business                167838.3211                   3/5/19   
3     New Business                418991.1971                   3/7/19   
4     New Business                418991.1971                   3/7/19   

  Opportunity_Close_Date Opportunity_Product  \
0                3/31/21  Privacy Monitoring   
1                6/30/21  Privacy Monitoring   
2                3/30/21  Privacy Monitoring   
3                4/16/21  Privacy Monitoring   
4                9/16/21  Privacy Monitoring   

  Account_Electronic_Medical_Record_System  ...  \
0                                    Other  ...   
1                                    Other  ...   
2                                    Other  ...   
3                                     Epic  ...   
4                                     Epic  ...   

  Account_Number_of_Affiliated_Physicians Account_Number_of_Employee  \
0                                   684.0               10236.916088   
1                                   684.0               10236.916088   
2                                   684.0               10236.916088   
3                                  3670.0               12464.000000   
4                                  3670.0               12464.000000   

  Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
0           21723248.0                   1.471406e+09   
1           21723248.0                   1.471406e+09   
2           21723248.0                   1.471406e+09   
3          160091963.0                   1.345325e+10   
4          160091963.0                   1.345325e+10   

   Account_Net_Patient_Revenue  Month  Year  Quarter  \
0                 3.705945e+08      9  2020       Q3   
1                 3.705945e+08     12  2020       Q4   
2                 3.705945e+08      3  2021       Q1   
3                 2.802628e+09      9  2020       Q3   
4                 2.802628e+09     12  2020       Q4   

   Opportunity_Annual_Amount_Prev_Quarter  Opportunity_Annual_Amount_Diff  
0                                     NaN                             NaN  
1                             167838.3211                             0.0  
2                             167838.3211                             0.0  
3                                     NaN                             NaN  
4                             418991.1971                             0.0  

[5 rows x 26 columns]

In [14]:
# Check what was that total 'Opportunity_Annual_Amount' for 'Closed Won' cases in 2023Q2
quarterly_df[(quarterly_df['Year_Quarter'] == '2023Q2') & (quarterly_df['Opportunity_Stage'] == 'Closed Won')]['Opportunity_Annual_Amount'].sum()

2442897.331

In [15]:
# Check what is the average 'Opportunity_Annual_Amount' that is not 'Closed Won' or 'Closed Lost' cases in 2023Q2
quarterly_df[(quarterly_df['Year_Quarter'] == '2023Q2') & (quarterly_df['Opportunity_Stage'] != 'Closed Won') & (quarterly_df['Opportunity_Stage'] != 'Closed Lost')]['Opportunity_Annual_Amount'].mean()

206723.75212697368

The total 'Fix' for both the effect of 'new appearing and closing deals' and the error in the value projection should amount to +9.5% .